Fail group install when package cannot be installed (#1461539)

AdamWill opened the pull request:
This should finally make DNF's behaviour when installing comps
groups the same as yum's used to be, which has been a request
for several years now. See:

https://bugzilla.redhat.com/show_bug.cgi?id=1292892
https://bugzilla.redhat.com/show_bug.cgi?id=1337731
https://bugzilla.redhat.com/show_bug.cgi?id=1427365
https://bugzilla.redhat.com/show_bug.cgi?id=1461539

And commits:

91f9ce98dbb630800017f44180d636af395435cd
1e1901822748b754d038dd396655c997281a7201

We have been around the houses on this multiple times. @mikem23
actually got things right way way back in #1292892:

"yum-deprecated behaves more sanely here.

In case 1 [package exists but has dependency issues], it reports
the missing dependency and exits with an error

...

In case 2 [package cannot be found at all], it exits cleanly, but
prints a warning."

Note that this applies to yum's *default* behaviour: it had an
option, --skip-broken, which caused it to skip packages with
dependency errors.

That's exactly what we've wanted all along, and it's also what
1461539 requests. Unfortunately, the initial commit intended to
fix #1292892 - that's 91f9ce9 - did not quite behave as Mike
requested. It treated any 'mandatory' package having dependency
errors *or* not existing at all as fatal errors, while not
treating non-existence *or* errors on the part of a 'default'
or 'optional' package as a fatal error. This introduced *two*
differences from yum's behaviour (making the comps 'type' matter,
which it never did to yum, and failing on non-existence), which
I think was the source of quite a lot of confusion.

We then wound up filing bugs on this - like #1337731 - and the
response to that was 1e19018, which causes dnf to treat *neither*
case as a fatal error for *any* package, which is where we stand
at present.

When resolving a transaction, currently, if it cannot find a
package corresponding to an entry in a group, dnf will log a
warning and carry on. This is good and fine and just what we want
it to do. However, when a package is available but cannot be
installed, it does the wrong thing. The various forks of this
`trans_install` internal function all wind up passing
`optional=True` to the Goal, which tells it that it's fine to
just leave the package out if its dependencies can't be resolved.
Additionally, if this happens, `resolve()` itself does not log
anything.

If you try a group install through the CLI, it will tell you
about packages that have been left out due to dependency errors,
via the `_skipped_packages` function in `dnf/output.py` which
actually sucks them out of the Goal instance itself. But that's
no use to anything besides the CLI. Other things which use
`resolve()`, like anaconda, have no means of accessing this
information, so when this happens to them, the omission of the
package is not logged in any way at all, and they have no way to
find out about it.

This commit simply inverts the behaviour for packages that do
exist but have dependency errors again, so if *any* package in
a groupinstall operation exists but cannot be installed, this is
a fatal error. That's what yum did, and that's what dnf should
have done all along. It continues to be the case that the comps
'type' - mandatory, default, or optional - does not matter; this
applies to all three. Again, this is how yum behaved (in the
absence of --skip-broken). If you said you want to install a group
and included optional packages, and any of those optional packages
had a dependency error, the transaction would fail.

The comps 'types' were *never* intended to dictate yum/dnf
behaviour in terms of whether it should fail on error or not, as I
understand it; they were in fact related to installer UI behaviour.
In the old installer UI, after you selected a group, you would see
all the packages in the group listed. 'mandatory' packages would
be pre-checked for installation and the check would be greyed out
- you couldn't unselect them. 'default' packages would be
pre-checked for installation, but you *could* de-select them if you
wanted. 'optional' packages would not be pre-checked for
installation, but you could *select* them if you wanted. This was
the intent of the comps 'type', as I understand it.

Importantly, a package not existing at all will continue to *not*
be a fatal error, as this is handled *before* we reach the
`trans_install` function.

Signed-off-by: Adam Williamson <awilliam@redhat.com>

AdamWill commented:
Excuse my tag farm, as this is quite a significant proposal that probably a few folks will want to know about. @ausil @nirik @mohanboddu @mluscon @M4rtinK @ignatenkobrain @dmach @mcepl @j-mracek @jkonecny12

Does anyone know folks from RHEL release engineering who would be appropriate to tag on this? I'm gonna guess they may have opinions.

The major practical consequence of this would be that compose processes which run through dnf (via lorax, anaconda or whatever) would once again *fail* if any package in an included comps group was available but could not be installed due to errors of some kind, as opposed to the package just being silently left out and the compose succeeding. Note that currently image composes fail if a package *specifically named in a kickstart* cannot be installed, but they do not fail on group packages not being installable.

AdamWill commented:
I think it's actually arguable that we should *also* make 'package does not exist' a fatal error again in future. However, before we can do that, we need to correct Fedora and probably RHEL comps not to list any non-existent packages in important groups. We would also need to change dnf and lorax to actually filter comps by arch: dnf has `base.read_comps(arch_filter=True)` for this, but it defaults to *False*, and while dnf's CLI sets it to true, neither anaconda nor lorax does.

AdamWill commented on dnf/base.py:
```diff
@@ -1451,15 +1451,15 @@ def trans_install(query, remove_query, comps_pkg):
                             _msg_installed(pkg)
                     sltr = dnf.selector.Selector(self.sack)
                     sltr.set(provides="({} if {})".format(comps_pkg.name, comps_pkg.requires))
-                    self._goal.install(select=sltr, optional=True)
```
This one I'm not *100%* sure about, actually. I *think* it'll still do the right thing regarding the conditional - i.e. it won't cause a fatal error if the 'depended-upon' package isn't selected for install - but it's possible I'm wrong, let me know if so.

j-mracek replied:
Here it will fail if no provides or impossible to install

j-mracek replied:
It will fail here even in case of unavailable provide.

j-mracek replied:
But it can be solved by query in advance.

AdamWill replied:
OK, I'll see if I can figure that out.

AdamWill replied:
sigh, this still needs figuring out, I think.

ausil commented:
@Tojaj FYI

ausil commented:
@lkocman @crungehottman FYI also

j-mracek commented:
@AdamWill This is not the first time when strict behavior of groups was implemented in DNF. Last time it was reverted because comps are permanently broken and due to request of comps description. Additionally PR does not make any difference if failing package is mandatory or something else. Simply I don't want to revert strict group behavior again.

lubomir commented:
> The major practical consequence of this would be that compose processes which run through dnf (via lorax, anaconda or whatever) would once again fail

I can't speak for everything, but at least for Pungi this change would not have any effect. It does not even use this functionality.

Tojaj commented:
> In case 1 [package exists but has dependency issues], it reports
the missing dependency and exits with an error
>
>...
>
>In case 2 [package cannot be found at all], it exits cleanly, but
prints a warning."

I like the proposed behavior.

Considering this is an important change in behavior, could be it implemented in phases?

Phase 1) The logic is implemented but only warnings are shown in both cases. This will last for few months at least. For case 1, the warning mentions that in future DNF version this will become an error. (The strict behavior could be enabled by a cmd option?)
Phase 2) In case 1DNF starts failing, case 2 still prints only warning

This way people would have time to properly test and prepare for the change.

lkocman commented:
This is ideally very wanted change, as RHEL-7 customers were often complaining that for whatever reason certain comps groups didn't not appear as installed.

Root cause for this behaviour was different than $TOPIC. 

But I see how "current" behaviour could lead to same question, no matter whether the notification that package was skipped was visible or not, since person checking list of installed groups might be different from the person seeing the warning.

Definitely up for failing the install completely.

AdamWill commented:
@lubomir not directly, no. But many of the tasks Pungi schedules are ultimately implemented via anaconda, and it is *anaconda's* behaviour that this would affect. Specific examples are live images and appliance disk images: currently, if the kickstarts for these specify comps groups, and packages from those comps groups have dependency problems, the image creation 'succeeds' with the packages silently dropped out. This is why we've had Workstation live images with gnome-software missing, and Cinnamon live images with...*cinnamon* missing.

AdamWill commented:
@j-mracek I specifically addressed all those points in the commit message, it's all written there.

The initial "strict" behaviour was implemented in response to [bug 1292892](https://bugzilla.redhat.com/show_bug.cgi?id=1292892), but it was *not* in fact the behaviour that bug requested. The key difference is that it caused dnf to fail *in the case that no package could be found for a 'mandatory' entry*, which is something yum *never did*. That's why it caused composes to fail and why we started complaining and filing bugs like [bug 1337731](https://bugzilla.redhat.com/show_bug.cgi?id=1337731) that led to the reversion.

The difference with this PR is that it implements - I believe - exactly yum's historic behaviour (with the difference that there is no `--skip-broken` override, though we could certainly put that back in, probably via the `strict` kwarg, if desired). That's what all the compose processes 'expect'. Specifically, with this PR, 'package cannot be found at all' is *not* a fatal error. *That* was the element of the initial 'strict' implementation that caused problems.

I also specifically explained why this PR *intentionally* does not care about the comps 'type': because yum never did, and *this is not what the 'type' is meant for*. The comps 'type' was ultimately all about the package selection UI in the old (pre-Fedora 18) anaconda. It was never intended to dictate how yum/dnf should behave in this kind of circumstance, that just isn't what it's for.

AdamWill commented:
@Tojaj That seems reasonable. Frankly the reason I didn't do it that way is it's quite a lot more complicated than just flipping the behaviour; like everything else in dnf this is achieved through about sixty-seven thousand levels of indirection, and the association with the `strict` kwarg got lost for comps group installs a couple of rewrites ago, so I'd have to figure out how to plumb that back in, then write tests for it all, and I was already kind of sick of this code by the time I finally figured out all the details of how it currently operates. :)

But I'll see if I can do something along those lines.

lkocman commented:
Just to say that our images use kickstart with --nobase option and most
packages are included explicitly by it's package name.

L

On Wed, Mar 14, 2018 at 4:29 PM, Adam Williamson <notifications@github.com>
wrote:

> @lubomir <https://github.com/lubomir> not directly, no. But many of the
> tasks Pungi schedules are ultimately implemented via anaconda, and it is
> *anaconda's* behaviour that this would affect. Specific examples are live
> images and appliance disk images: currently, if the kickstarts for these
> specify comps groups, and packages from those comps groups have dependency
> problems, the image creation 'succeeds' with the packages silently dropped
> out. This is why we've had Workstation live images with gnome-software
> missing, and Cinnamon live images with...*cinnamon* missing.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rpm-software-management/dnf/pull/1038#issuecomment-373063179>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAfIp9sM1KrXksILWsj6frJBt2ihm0oIks5teTdYgaJpZM4Sptex>
> .
>

j-mracek commented on dnf/base.py:
```diff
@@ -1442,7 +1442,7 @@ def trans_upgrade(query, remove_query, comps_pkg):
             if self.conf.multilib_policy == "all":
                 if not comps_pkg.requires:
-                    self._install_multiarch(query, strict=False)
+                    self._install_multiarch(query, strict=True)
```
In this case strict here should be an inherited variable from group_install function.

AdamWill replied:
yes, I figured that's what was going on there, but the association was lost in a refactor a while back, and isn't super straightforward to restore, as these are distinct operations - we'd have to find somewhere appropriate to stash the `strict` setting between the `group_install` and the `resolve`. Really, this whole `strict` mechanism feels somewhat wobbly to me - it's never clear whether `strict` is supposed to be a property of the overall DNF configuration, or this specific `Base` instance's configuration, or a transaction, or some specific operation that feeds into a transaction...the contexts seem to get mixed up a lot. Unless I'm missing something.

AdamWill commented:
@j-mracek responding to something from IRC...you wrote:

(jmracek) According what I have seen - @Trojaj Phase 1 is a present behavior of DNF. Missing packages are reported, as warning. Skipped packages are reported in transaction table.

This is true but *only of the CLI*.

The logging of non-existent packages is fine, actually. That's implemented as a real DNF log message ultimately within `resolve()`, so users of the DNF API do receive the log message.

However, the logging of skipped packages is not. It's done *only for the CLI* and relies on a private attribute of the base instance. The code is [here](https://github.com/rpm-software-management/dnf/blob/003e278b1ad2672734903f5f05216dbe21b12e79/dnf/cli/output.py#L1030-L1043) - note that this is within the CLI code, and relies on the private `self.base._goal` attribute. It essentially clones and re-runs the goal (which is a hawkey/libdnf `Goal` instance) and then pulls out the information about skipped packages from the goal. That works for the CLI, but it's no use at all to anything else - specifically it's no use to anaconda, which is using the DNF API. When anaconda calls `resolve()`, it does not get this information at all. It can't even correctly copy what the CLI does, because of `base._goal` being a private attribute. There is just no public interface through which a DNF API consumer could properly access this information, so far as I can figure out.

M4rtinK commented:
I like this change - the current situation is kinda similar to silent disk corruption (some packages just kinda get lost at random - oops) and that's definitely something to fix.

I've checked the [documentation for the %packages selection](https://pykickstart.readthedocs.io/en/latest/kickstart-docs.html#chapter-6-package-selection) in kickstart and it looks like we don't promise any guarantees this change would compromise. The closest is the `--ignoremissing` option, but that corresponds only to groups and packages listed by name in kickstart, it does not say anything about packages installed by the groups named in kickstart. So this change so far looks clean from the point of view of kickstart backwards compatibility.

BTW, I was wondering - would it be possible automatically check for any missing packages before the compose process is even started ? The input data should be available (default groups, comps, repository & kickstart) and I don't think there is anything special in Anaconda that would be needed to just check all packages requested via groups are available and look installable (exist, don't conflict & deps are satisfied, etc.).

Of course this would not be able to uncover things like failing scriptlets, but that's not what this change is about anyway if I understand things correctly.

AdamWill commented:
@M4rtinK possible, sure. In fact I believe @dustymabe is working on something along those lines. I only worry that it involves reimplementing some stuff that's probably already implemented more times than it needs to be, but hey. It's a bit off-topic for this PR, though.

For the record I am thinking about how to update this PR to include some configurability, in a way that makes as much sense as possible. F28 Beta is pretty busy right now, though.

dustymabe commented:
> @M4rtinK possible, sure. In fact I believe @dustymabe is working on something along those lines.

Description of possible workflow for detecting this is documented in releng issue [#7388](https://pagure.io/releng/issue/7388). Please add input there, or feel free to pick it up if you think you would like to write the tool.

AdamWill commented:
This sideci thing seems highly toxic to me: refuses to show me results unless I "sign in", compulsorily via github. i.e., let them datamine my github profile. I'm not doing that. So I can't see the detailed results of the test execution, I'm just going to have to assume the rebase worked as intended.

I'm intending to try and revise this to be more configurable now.

AdamWill commented:
OK, there's an initial cut at making this behaviour somewhat configurable. I chose to do it the most minimal way I could think of, which was to just wire up the existing `strict` args again. I elected to do this by restoring the `install` / `install_opt` separation that was lost in 1e19018 , that turned out not to be terribly hard.

Note this does change the 'default' behaviour back to the yum-like "strict" implementation, since the default values of the `strict` args to both `group_install` and `env_group_install` are and always have been `True`. Changing this is actually slightly tricky as it's hard to know what to do with the `env_group_install` case - we could just change the `group_install` default to `False`, that's easy enough, but `env_group_install` would still default to strict `group_install` behaviour in that case, and just changing `env_group_install`'s default to `False` as well would actually change its own behaviour in the case there's nothing to do (it would not raise an error by default any more, whereas at present it does).

The [potential issue](https://github.com/rpm-software-management/dnf/pull/1038#pullrequestreview-103674376) with conditional entries in comps still needs to be figured out, too.

AdamWill commented:
So, here's current state: I've got a local branch with the same basic changes cherry-picked on top of the `dnf-2-modularity` branch (which requires quite a lot of change unfortunately, as various relevant bits are different on those two branches) and I've been running the tests on that branch, as I don't have sufficiently new libdnf/hawkey locally to run the tests on this branch. The tests pass now, which *should* mean that the 'conditional' case is acting as we intend - even with `strict=True`, 'dependency for a conditional package is not available' is not a critical error. However, this code path looks different on `dnf-2-modularity` vs. `master` - on `dnf-2-modularity` the `conditional_pkg_check` nested method is used, on `master` things were changed by cca691818 , so things may behave differently. I'll see if I can set up a container or VM or something with new enough libdnf bits to run the tests on this branch later.

AdamWill commented:
Further update: I tweaked master a bit, and managed to get a VM with sufficiently new bits for the tests to run. In this state the tests pass. Which suggests the 'conditional' behaviour is correct, unless I messed up the tests.

j-mracek commented:
During of testing of your patch I have discovered additional issue in upstream. I am looking for solution then I will finish the review.

AdamWill commented:
It appears that a major reimplementation of all this stuff was done recently and is currently landing in Rawhide...which *once again* entirely ignored all of this stuff that I've been carefully explaining for months. :(

j-mracek commented:
Thank, it works like expected. The problem that I mentioned is handled already before reach dep is formatted.
